contrib/fparser/fparser_ad.cc

```diff
@@ -111,6 +114,95 @@ Value_t FunctionParserADBase<Value_t>::fp_plog(const Value_t * params)
   return x < a ? fp_log(a) + (x-a)/a - (x-a)*(x-a)/(Value_t(2)*a*a) + (x-a)*(x-a)*(x-a)/(Value_t(3)*a*a*a) : fp_log(x);
 }
 
+template <typename Value_t>
+Value_t FunctionParserADBase<Value_t>::fp_erf(const Value_t * params)
+{
+  const Value_t x = params[0];
+  return std::erf(x);
+}
+
+// Derivation of these complex method can be found here: https://math.stackexchange.com/a/712568/408963
+
+template<>
+std::complex<double>
+FunctionParserADBase<std::complex<double>>::fp_erf(const std::complex<double> * params)
+{
+  const std::complex<double> z = params[0];
+  double x = z.real(), y = z.imag();
+  double fx = std::erf(x) + std::exp(-(x*x)) / (2. * M_PI) * (1. - std::cos(2. * x * y));
+  double fy = std::exp(-(x*x)) / (2. * M_PI) * std::sin(2. * x * y);
+
+  // Set this to as high as you want for greater accuracy
+  unsigned n = 3;
+  for (unsigned k = 1; k <= n; k++)
+  {
+    double fxk = 2. * x * (1. - std::cos(2. * x * y) * std::cosh(k * y))
+      + k * std::sin(2. * x * y) * std::sinh(k * y);
+    double fyk = 2. * x * std::sin(2. * x * y) * std::cosh(k * y)
+      + k * std::cos(2. * x * y) * std::sinh(k * y);
+    double mult = 2. / M_PI * std::exp(-(x*x)) * std::exp(-k*k / 4.) / (k*k + 4. * x*x);
+    fx += mult * fxk;
+    fy += mult * fyk;
+  }
+
+  std::complex<double> f (fx, fy);
+  return f;
+}
+
+template<>
+std::complex<float>
+FunctionParserADBase<std::complex<float>>::fp_erf(const std::complex<float> * params)
+{
+  const std::complex<float> z = params[0];
+  float x = z.real(), y = z.imag();
+  double fx = std::erf(x) + std::exp(-(x*x)) / (2. * M_PI) * (1. - std::cos(2. * x * y));
+  double fy = std::exp(-(x*x)) / (2. * M_PI) * std::sin(2. * x * y);
+
+  // Set this to as high as you want for greater accuracy
+  unsigned n = 3;
+  for (unsigned k = 1; k <= n; k++)
+  {
+    double fxk = 2. * x * (1. - std::cos(2. * x * y) * std::cosh(k * y))
+      + k * std::sin(2. * x * y) * std::sinh(k * y);
+    double fyk = 2. * x * std::sin(2. * x * y) * std::cosh(k * y)
+      + k * std::cos(2. * x * y) * std::sinh(k * y);
+    double mult = 2. / M_PI * std::exp(-(x*x)) * std::exp(-k*k / 4.) / (k*k + 4. * x*x);
+    fx += mult * fxk;
+    fy += mult * fyk;
+  }
+
+  std::complex<float> f (fx, fy);
+  return f;
+}
+
+template<>
+std::complex<long double>
```

Comment:
It was those instantiations that led me to code the complex method specializations in the first place. (Out of curiosity, do we actually have people using function parser with complex numbers?). If I comment out the `long double` and `float` specializations, then I get the error:
```
../../../contrib/fparser/fparser_ad.cc:123:10: error: no matching function for call to 'erf'
  return std::erf(x);
         ^~~~~~~~
../../../contrib/fparser/fparser_ad.cc:1203:16: note: in instantiation of member function 'FunctionParserADBase<std::complex<long double> >::fp_erf' requested here
template class FunctionParserADBase< std::complex<long double> >; template class ADImplementation< std::complex<long double> >;
               ^
/usr/include/x86_64-linux-gnu/bits/mathcalls.h:259:15: note: candidate function not viable: no known conversion from 'const std::complex<long double>' to 'double' for 1st argument
extern double erf (double) throw (); extern double __erf (double) throw ();
              ^
/opt/moose/gcc-6.2.0/lib/gcc/x86_64-pc-linux-gnu/6.2.0/../../../../include/c++/6.2.0/cmath:1306:3: note: candidate function not viable: no known conversion from 'const std::complex<long double>' to 'float' for 1st argument
  erf(float __x)
  ^
/opt/moose/gcc-6.2.0/lib/gcc/x86_64-pc-linux-gnu/6.2.0/../../../../include/c++/6.2.0/cmath:1310:3: note: candidate function not viable: no known conversion from 'const std::complex<long double>' to 'long double' for 1st argument
  erf(long double __x)
  ^
/opt/moose/gcc-6.2.0/lib/gcc/x86_64-pc-linux-gnu/6.2.0/../../../../include/c++/6.2.0/cmath:1317:5: note: candidate template ignored: substitution failure [with _Tp = std::complex<long double>]: no type named '__type' in '__gnu_cxx::__enable_if<false, double>'
    erf(_Tp __x)
```